### PR TITLE
Add property to disable problem report generation

### PR DESCRIPTION
**Pull request by reinsch82**
Fixes https://github.com/gradle/gradle-private/issues/4478

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonWriter.kt [hunk 1]
```diff
@@ -20,13 +20,24 @@ import com.fasterxml.jackson.core.JsonFactory
...
+    class JsonObject(val jsonWriter: JsonWriter) : AutoCloseable {
+        init {
+            jsonWriter.beginObject()
+        }
+
+        override fun close() {
+            jsonWriter.endObject()
+        }
+    }
+
```
Comment (donat) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonWriter.kt:33]: ❌ I don't have context of why this change is required in this PR.

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonWriter.kt [hunk 2]
```diff
@@ -20,13 +20,24 @@ import com.fasterxml.jackson.core.JsonFactory
...
     private
     val jsonGenerator = JsonFactory().createGenerator(writer)
 
     fun jsonObject(body: () -> Unit) {
-        beginObject()
-        body()
-        endObject()
+        JsonObject(this).use {
```
Comment (donat) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonWriter.kt:38]: Same as above

File: platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy [hunk 3]
```diff
@@ -63,13 +63,13 @@ abstract class AbstractBasicGroupedTaskLoggingFunctionalTest extends AbstractCon
         result.groupedOutput.taskCount == 3
         if (errorsShouldAppearOnStdout()) {
             // both stdout and stderr are attached to the console
-            assert result.groupedOutput.task(':1:log').output == "Error from 1\nOutput from 1\nDone with 1\nDone with 1"
-            assert result.groupedOutput.task(':2:log').output == "Error from 2\nOutput from 2\nDone with 2\nDone with 2"
-            assert result.groupedOutput.task(':3:log').output == "Error from 3\nOutput from 3\nDone with 3\nDone with 3"
+            assert result.groupedOutput.task(':1:log').output.startsWith("Error from 1\nOutput from 1\nDone with 1\nDone with 1")
```
Comment (donat) [inline platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy:66]: Same as above

File: platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy [hunk 4]
```diff
@@ -435,12 +424,42 @@ class ProblemsServiceIntegrationTest extends AbstractIntegrationSpec {
...
+            }
+        """
+
+        executer.withArguments("--disable-problem-report")
 
         when:
         run("reportProblem")
 
         then:
+        !testDirectory.file("build/reports/problems", "problem-report.html").exists()
```
Comment (donat) [inline platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy:462]: ❌ Missing assertion for console output not having the link to the report.

File: subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java [hunk 5]
```diff
@@ -649,4 +647,19 @@ public void applyTo(StartParameterInternal settings, Origin origin) {
             settings.setPropertyUpgradeReportEnabled(true);
         }
     }
+
+    public static class ProblemReportGenerationOption extends EnabledOnlyBooleanBuildOption<StartParameterInternal> {
+
+        public static final String LONG_OPTION = "disable-problem-report";
+        public static final String GRADLE_PROPERTY = "org.gradle.problem.report.disabled";
```
Comment (donat) [inline subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java:654]: ❌ I think we can improve the property a little bit to fit better with the existing gradle properties.  I propose following:
- let's not use negative terms
- the property key should be org.gradle.problem.report with possible values `enabled` and `disabled` (or `true` and `false)
- the default value should be set to true
- we should have two command-line options: `--problem-report` and `--no-problem-report`.

File: subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java [hunk 6]
```diff
@@ -649,4 +647,19 @@ public void applyTo(StartParameterInternal settings, Origin origin) {
...
+
+        public ProblemReportGenerationOption() {
+            super(GRADLE_PROPERTY, CommandLineOptionConfiguration.create(LONG_OPTION, "(Experimental) disables HTML problem report generation"));
+        }
```
Comment (donat) [inline subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java:658]: 🤔 Maybe the URL we show on the console should be alsow prefixed with `incubating:`

File: subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java [hunk 7]
```diff
@@ -649,4 +647,19 @@ public void applyTo(StartParameterInternal settings, Origin origin) {
...
+
```
Comment (donat) [inline subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java:659]: Please don't forget the release notes.

**Review by donat — CHANGES_REQUESTED** [review]

File: subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java (hunk 7 above)
Comment (donat) [reply subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java:659]: Or at least a placeholder

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonWriter.kt (hunk 2 above)
Comment (reinsch82) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonWriter.kt:38]: some test broke because of an invalid state during JSON generation.
this guarantees a valid json for this cases.

File: platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy (hunk 3 above)
Comment (reinsch82) [reply platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractBasicGroupedTaskLoggingFunctionalTest.groovy:66]: startsWith is a mitigation because the "--no-problem-report" doesn't work here.
so I hope I can fix it with some feedback soon.

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractConsoleBuildResultFunctionalTest.groovy [hunk 8]
```diff
@@ -89,7 +89,8 @@ BUILD SUCCESSFUL in [ \\dms]+
         """
 
         when:
-        succeeds('success')
+        args "--no-problems-report"
```
Comment (donat) [inline platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractConsoleBuildResultFunctionalTest.groovy:92]: ❓ Can we add this to `AbstractIntegrationSpec.setup()`?

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt [hunk 9]
```diff
@@ -74,15 +74,15 @@ class DefaultProblemsReportCreator(
...
+                            property("documentationLink", DocumentationRegistry().getDocumentationFor("problems-report"))
                             property("documentationLinkCaption", "Problem report")
                         }
                     }
                 }
             }
         })?.let {
             val url = ConsoleRenderer().asClickableFileUrl(it)
-            logger.warn("Problems report is available at: $url")
+            logger.warn("Problem report (Experimental) is available at: $url")
```
Comment (donat) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt:85]: 🤔 

I'd suggest to add a prefix instead

    logger.warn("[Experimental] Problem report is available at: $url")

or

    logger.warn("[Incubating] Problem report is available at: $url")

File: platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractConsoleBuildResultFunctionalTest.groovy (hunk 8 above)
Comment (donat) [inline platforms/core-runtime/logging/src/integTest/groovy/org/gradle/internal/logging/console/taskgrouping/AbstractConsoleBuildResultFunctionalTest.groovy:92]: ❌ Also, we should have a follow-up issue for maintenance day to get rid of the now-added `--no-problems-report` flags.

File: subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java [hunk 10]
```diff
@@ -36,56 +37,96 @@
...
+//        options.add(new ConfigurationCacheIgnoredFileSystemCheckInputs());
+//        options.add(new ConfigurationCacheDebugOption());
+//        options.add(new ConfigurationCacheParallelOption());
+//        options.add(new ConfigurationCacheRecreateOption());
+//        options.add(new ConfigurationCacheQuietOption());
+//        options.add(new IsolatedProjectsOption());
+//        options.add(new ProblemReportGenerationOption());
+//        options.add(new PropertyUpgradeReportOption());
+//        StartParameterBuildOptions.options = Collections.unmodifiableList(options);
+//    }
```
Comment (donat) [inline subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptions.java:129]: ❌ Leftover comment should be removed.

```suggestion
```

**Review by donat — CHANGES_REQUESTED** [review]

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/documentation/docs/src/docs/release/notes.md [hunk 11]
```diff
@@ -128,6 +128,14 @@ If only warnings happen (and no `-Werror` is set), this report will not be visib
 
 This feature also works with the [`--continue`](userguide/command_line_interface.html#sec:continue_build_on_failure) flag, and the bottom report will contain a per-task report of all the compilation failures.
 
+#### HTML problems report
+
+All problems reported via the [Problems API](userguide/implementing_gradle_plugins_binary.html#reporting_problems), internally in Gradle or from a plugin, will be rendered into an HTML problems report
+by default.
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:134]: Is it possible to show an example console output that showcases the HTML report?

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: testing/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy [hunk 12]
```diff
@@ -239,7 +239,10 @@ abstract class DistributionIntegrationSpec extends AbstractIntegrationSpec {
 
     protected void checkMinimalContents(TestFile contentsDir) {
         // Check it can be executed
-        executer.inDirectory(contentsDir).usingExecutable('bin/gradle').withTasks("help").run()
+
+        def directory = executer.inDirectory(contentsDir)
+//        if(directory.distribution.version
```
Comment (donat) [inline testing/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy:244]: ❌ 

```suggestion
//        if(directory.distribution.version
```

File: platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenSnapshotResolveIntegrationTest.groovy [hunk 13]
```diff
@@ -972,7 +980,7 @@ dependencies {
     compile "group:projectA:1.0"
 }
 """
-        run 'retrieve'
+        runRetrieve()
```
Comment (donat) [inline platforms/software/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/maven/MavenSnapshotResolveIntegrationTest.groovy:983]: 🤔 

```suggestion
        runRetrieveTask()
```

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt [hunk 14]
```diff
@@ -74,15 +74,15 @@ class DefaultProblemsReportCreator(
...
+                            property("documentationLink", DocumentationRegistry().getDocumentationFor("problems-report"))
                             property("documentationLinkCaption", "Problem report")
                         }
                     }
                 }
             }
         })?.let {
             val url = ConsoleRenderer().asClickableFileUrl(it)
-            logger.warn("Problems report is available at: $url")
+            logger.warn("[Incubating] Problem report is available at: $url")
```
Comment (donat) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt:85]: ❌ 
```suggestion
            logger.warn("[Incubating] Problems report is available at: $url")
```

File: platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy (hunk 4 above)
Comment (donat) [reply platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy:420]: This is still missing

**Review by donat — CHANGES_REQUESTED** [review]

**Comment by reinsch82** [comment]
@donat please review, also ok after the merge.
